### PR TITLE
`String.prototype.isWellFormed`/`String.prototype.toWellFormed`: add tests

### PR DESCRIPTION
**Pull request by ljharb**
Tests for the stage 2 proposal https://github.com/tc39/proposal-is-usv-string.

### CODE REVIEW

File: test/built-ins/String/prototype/isWellFormed/returns-boolean.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  2. Let S be ? ToString(O).
+  3. Return IsStringWellFormedUnicode(S).
+
+features: [String.prototype.isWellFormed]
+---*/
+assert.sameValue(typeof String.prototype.isWellFormed, 'function');
+
+assert.sameValue('\uD800'.isWellFormed(), false, 'a string with a lone surrogate is not well-formed');
+assert.sameValue('💩'.isWellFormed(), true, 'a surrogate pair is well-formed');
+assert.sameValue('abc'.isWellFormed(), true, 'a normal string is well-formed');
```
Comment (michaelficarra) [inline test/built-ins/String/prototype/isWellFormed/returns-boolean.js:20]: The cases I would test:

* a lead surrogate surrounded by other non-surrogates
* a trail surrogate surrounded by other non-surrogates
* adjacent lead/trail surrogates but in the wrong order
* a well-formed string constructed by concatenating a string ending with a lead surrogate and a string starting with a trail surrogate
* a non-well-formed string constructed by slicing a well-formed string in the middle of a surrogate pair
* a string with non-BMP codepoints but no surrogates
* a well-formed string with surrogates
* a latin-1 (7-bit ASCII) string

Comment (michaelficarra) [inline test/built-ins/String/prototype/isWellFormed/returns-boolean.js:20]:
```suggestion
assert.sameValue('abc'.isWellFormed(), true, 'a latin-1 string is well-formed');
```

I'd refrain from using terms like "normal" when it comes to culturally significant things such as language.

Comment (ljharb) [reply test/built-ins/String/prototype/isWellFormed/returns-boolean.js:20]: I'm not sure how the string construction method makes a difference?

Comment (ljharb) [reply test/built-ins/String/prototype/isWellFormed/returns-boolean.js:20]: Also, can you suggest some non-surrogate non-BMP codepoints to use?

Comment (michaelficarra) [reply test/built-ins/String/prototype/isWellFormed/returns-boolean.js]: The reason it's important to include tests for a variety of ways that a string can come to be is that it is likely that an implementation trying to provide sub-linear results will need to keep a flag associated with a string. This flag needs to be propagated appropriately across concatenation and slicing. These tests will help ensure that they've done that.

> Also, can you suggest some non-surrogate non-BMP codepoints to use?

Oops, I meant non-ASCII, so anything else in the BMP. Sorry. How about U+25A8 ▨ (SQUARE WITH UPPER RIGHT TO LOWER LEFT FILL)? Also, side note: why do we have a geometric shapes block in the BMP?!

File: test/built-ins/String/prototype/isWellFormed/returns-boolean.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  false,
+  'leading lone surrogates are not well-formed'
+);
+assert.sameValue(
+  ('a' + trailingPoo + 'c' + trailingPoo + 'e').isWellFormed(),
+  false,
+  'trailing lone surrogates are not well-formed'
+);
+assert.sameValue(
+  ('a' + leadingPoo + trailingPoo + 'd').isWellFormed(),
```
Comment (michaelficarra) [inline test/built-ins/String/prototype/isWellFormed/returns-boolean.js:33]: This is backwards. Well, not backwards. You know what I mean.

Comment (ljharb) [reply test/built-ins/String/prototype/isWellFormed/returns-boolean.js:33]: whoops, on it

**Review by michaelficarra — APPROVED** [review]

File: test/built-ins/String/prototype/toWellFormed/to-string.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// Copyright (C) 2022 Jordan Harband. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-String.prototype.isWellFormed
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/toWellFormed/to-string.js:4]: Why is this in the `toWellFormed` folder?

File: test/built-ins/String/prototype/isWellFormed/length.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+assert.sameValue(typeof String.prototype.isWellFormed, 'function');
+
+assert.sameValue(
+  String.prototype.isWellFormed.length, 0,
+  'The value of String.prototype.isWellFormed.length is 0'
+);
+
+verifyNotEnumerable(String.prototype.isWellFormed, 'length');
+verifyNotWritable(String.prototype.isWellFormed, 'length');
+verifyConfigurable(String.prototype.isWellFormed, 'length');
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/isWellFormed/length.js:24]: Suggest using `verifyProperty` instead.

File: test/built-ins/String/prototype/isWellFormed/not-a-constructor.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  isConstructor(String.prototype.isWellFormed),
+  false,
+  'isConstructor(String.prototype.isWellFormed) must return false'
+);
+
+assert.throws(TypeError, function () {
+  new String.prototype.isWellFormed();
+}, '`new String.prototype.isWellFormed()` throws TypeError');
+
+
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/isWellFormed/not-a-constructor.js:34]: Please check all tests for empty lines at the end of the file.

File: test/built-ins/String/prototype/isWellFormed/prop-desc.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+  17 ECMAScript Standard Built-in Objects
+includes: [propertyHelper.js]
+features: [String.prototype.isWellFormed]
+---*/
+assert.sameValue(typeof String.prototype.isWellFormed, 'function');
+
+assert.sameValue(
+  typeof String.prototype.isWellFormed,
+  'function',
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/isWellFormed/prop-desc.js:18]: Duplicate test

File: test/built-ins/String/prototype/isWellFormed/return-abrupt-from-this.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+// Copyright (C) 2022 Jordan Harband. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-string.prototype.is-well-formed
+description: >
+  Return abrupt from RequireObjectCoercible(this value).
+info: |
+  String.prototype.isWellFormed( )
+
+  Let O be ? RequireObjectCoercible(this value).
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/isWellFormed/return-abrupt-from-this.js:10]:
```suggestion
  1. Let O be ? RequireObjectCoercible(this value).
```

File: test/built-ins/String/prototype/isWellFormed/returns-boolean.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
+// Copyright (C) 2022 Jordan Harband. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-String.prototype.at
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/isWellFormed/returns-boolean.js:4]: Not sure that's right

File: test/built-ins/String/prototype/isWellFormed/returns-boolean.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+description: >
+  The method should return an Iterator instance.
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/isWellFormed/returns-boolean.js:6]: Not sure that's right either

File: test/built-ins/String/prototype/isWellFormed/returns-boolean.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+assert.sameValue(typeof String.prototype.isWellFormed, 'function');
+
+var leadingPoo = '\uD83D';
+var trailingPoo = '\uDCA9';
+var wholePoo = leadingPoo + trailingPoo;
+
+assert.sameValue(
+  ('a' + leadingPoo + 'c' + leadingPoo + 'e').isWellFormed(),
+  false,
+  'leading lone surrogates are not well-formed'
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/isWellFormed/returns-boolean.js:25]: Should be "lone leading surrogates", I think. Same below

File: test/built-ins/String/prototype/isWellFormed/to-string.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
+// Copyright (C) 2022 Jordan Harband. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-String.prototype.isWellFormed
+description: >
+  The method should return an Iterator instance.
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/isWellFormed/to-string.js:6]: Same

File: test/built-ins/String/prototype/toWellFormed/length.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+assert.sameValue(typeof String.prototype.toWellFormed, 'function');
+
+assert.sameValue(
+  String.prototype.toWellFormed.length, 0,
+  'The value of String.prototype.toWellFormed.length is 0'
+);
+
+verifyNotEnumerable(String.prototype.toWellFormed, 'length');
+verifyNotWritable(String.prototype.toWellFormed, 'length');
+verifyConfigurable(String.prototype.toWellFormed, 'length');
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/toWellFormed/length.js:24]: verifyProperty

File: test/built-ins/String/prototype/toWellFormed/not-a-constructor.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  implement the [[Construct]] internal method unless otherwise specified in
+  the description of a particular function.
+
+  sec-evaluatenew
+
+  ...
+  7. If IsConstructor(constructor) is false, throw a TypeError exception.
+  ...
+includes: [isConstructor.js]
+features: [Reflect.construct]
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/toWellFormed/not-a-constructor.js:21]: Missing feature

File: test/built-ins/String/prototype/toWellFormed/length.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
+// Copyright (C) 2022 Jordan Harband. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-string.prototype.to-well-formed
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/toWellFormed/length.js:4]: I don't think this id should have dashes in it

File: test/built-ins/String/prototype/toWellFormed/prop-desc.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
+// Copyright (C) 2020 Rick Waldron. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-string.prototype.is-well-formed
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/toWellFormed/prop-desc.js:4]: "is" → "to"

File: test/built-ins/String/prototype/toWellFormed/prop-desc.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
+// Copyright (C) 2020 Rick Waldron. All rights reserved.
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/toWellFormed/prop-desc.js:1]: Is that right?

File: test/built-ins/String/prototype/toWellFormed/returns-well-formed-string.js [hunk 17]
```diff
@@ -0,0 +1,55 @@
+// Copyright (C) 2022 Jordan Harband. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: sec-String.prototype.at
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/toWellFormed/returns-well-formed-string.js:4]: Ahem

File: test/built-ins/String/prototype/toWellFormed/to-string.js [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+description: >
+  The method should return an Iterator instance.
```
Comment (Ms2ger) [inline test/built-ins/String/prototype/toWellFormed/to-string.js:6]: Ahem

**Review by Ms2ger — CHANGES_REQUESTED** [review]

File: test/built-ins/String/prototype/toWellFormed/to-string.js (hunk 3 above)
Comment (ljharb) [reply test/built-ins/String/prototype/toWellFormed/to-string.js:4]: copy-paste error, thanks, will fix

File: test/built-ins/String/prototype/isWellFormed/length.js (hunk 4 above)
Comment (ljharb) [reply test/built-ins/String/prototype/isWellFormed/length.js:24]: I copied files initially from String.prototype.at; I’ll update this.

File: test/built-ins/String/prototype/isWellFormed/prop-desc.js (hunk 6 above)
Comment (ljharb) [reply test/built-ins/String/prototype/isWellFormed/prop-desc.js:18]: actually prop-desc seems the appropriate place to test the type of the value.

**Review by ptomato — APPROVED** [review]
This proposal reached stage 3 today and this PR has had 2 reviews, one from a maintainer and one from the proposal champion. I think it's time to merge it, so this is a rubberstamp.

The linter seems to be complaining about a couple of files that don't have newlines at the end (https://app.circleci.com/pipelines/github/tc39/test262/4566/workflows/c2330ae2-25e9-4ecc-b925-4965328f5925/jobs/42689) but I think this should be all clear to merge after fixing that and rebasing.

**Pull request merged**